db/document.go

```diff
@@ -829,6 +832,23 @@ func (doc *Document) UpdateExpiry(expiry uint32) {
 
 //////// CHANNELS & ACCESS:
 
+func (doc *Document) updateChannelHistory(channelName string, seq uint64, addition bool) {
+	for idx, chanName := range doc.ChanNames {
+		if chanName == channelName {
+			if !addition {
+				doc.ChannelHistory[idx][len(doc.ChannelHistory[idx])-1].EndSeq = seq
+			} else {
+				doc.ChannelHistory[idx] = append(doc.ChannelHistory[idx], auth.GrantHistorySequencePair{StartSeq: seq})
+			}
+			return
+		}
+	}
+
+	// First time addition
+	doc.ChanNames = append(doc.ChanNames, channelName)
```

Comment:
Needs changing. This assumes that if there is no entry its going to be startSeq entry. In reality this could be removal of a document from before upgrade. Here we can check the addition bool and set start as 1 and add end if its a removal.

Reply:
We'll need to think about when to generate the new format for legacy documents.